### DISCUSSION TITLE
Tag system for files / directories

### DISCUSSION THREAD

**Discussion by pixel-boy** [godotengine/godot-proposals#10562] [Editor]
Be able to define a list of tags in ProjectSettings which can be assigned to files ( scenes, resources and probably external file like PNG,WAV etc... ) and directories.
![tag_preview](https://github.com/user-attachments/assets/8447a474-d1f4-41b7-8cc3-651affd9caa1)

### 1 - Better navigation in the file system:
Navigating in big projects is tedious, the "favorite" option is supposed to help with this, but I've never found it practical, and I think a tag system could be an improved version of it (it would be possible to have default tags like "favorite" "debug" etc...)

Another example in the case of level creation, it would be possible to create an "LD" tag to easily find all the level design elements.

### 2 - Get file lists via code

It's often for debug features that I feel the need for it, but being able to get the list of all the files of a resource type (ex: ResourceItem) or all the scenes that are of a certain class (ex: Map) would be nice, and the tag system could allow to do this.

### Create tags
![tag_project_settings](https://github.com/user-attachments/assets/9c847255-746c-4efa-a35a-8dda76ab3812)

In the file system, tags could appear discreetly to the right of the files (the complete list could be displayed on hover)
![tag_in_file_system](https://github.com/user-attachments/assets/4b343eef-1855-4e05-9e96-7d3edd42ec5c)

### Assigning tags to files / directories

- Via the interface in the file system
![set_tag](https://github.com/user-attachments/assets/b969e81e-7951-417c-a11e-35dddea67b42)

- With annotation for custom classes
![tag annotation](https://github.com/user-attachments/assets/bd76196c-9c7b-4486-b3fc-5febc56234bf)

- And via code ( maybe with DirAccess / FileAccess ? )
`DirAccess.add_tag(file_path,tag)`
`DirAccess.remove_tag (file_path,tag)`

#### In file system
![filter by tag](https://github.com/user-attachments/assets/c17ac4e6-e857-4e38-9b1a-ce9ea3d31e59)

#### With code
![get_files_by_tag](https://github.com/user-attachments/assets/7e07c342-9f2b-4010-9a27-6be561e77fd3)

### Similar discussion
https://github.com/godotengine/godot-proposals/discussions/10457 This proposal is not the same but offers another way to improve navigation in the file system

**Comment by kaiiboraka** [comment]
Oh, yes. Please, please, please. Also, I love your mockups. Nicely done.
I just want to add some extra thoughts for why I believe this would be so useful, at least for me.

As with most game projects, our files ultimately have a 2-Dimensional structure (at least). Where on the one hand, I have different Mediums of content/File Types, categorizing by what the file itself _is_, such as Scenes, Scripts, Art Assets, Audio Assets, and so on. Then there is another dimension where we categorize by subject matter association, where it's based on what the file _contains_ or _is about_, such as Player, Enemy, Level, Environment, Objects.

Currently, we are FORCED to choose ONE of those to divide everything by, the other of which will be redundantly reincluded in EVERY SINGLE ONE of those top-level folders. So if I did it by File Type, I would have: Scripts/ with subfolders Player, Enemy, Level, etc.; Scenes/ with subfolders Player, Enemy, Level, etc.; Art/ with subfolders Player, Enemy, Level, etc. it gets old fast.
Then there's the other way: Player/ Scripts, Scenes, Art, Audio, etc.; Enemies/ Scripts, Scenes, Art, Audio, etc.; Levels/ Scripts, Scenes, Art, Audio, etc.;

There's no complete answer to this problem without breaking out of the inherent limitations of a linear hierarchichal file structure, and the only way to do that is to either allow items to exist in more than one place such as with symbolic links, or to at least be able to FIND any arbitrary item from where you currently are via tag filtering. While I'm at it, cluttering file names with descriptors is a horrible way to "tag" things...

Semi-relatedly, there exists [a plugin for Obsidian called tagfolder](https://github.com/vrtmrz/obsidian-tagfolder), and it is incredible. It removes the barrier entirely, by creating a "hierarchical" structure out of composing tags together as filters. It lets you describe what something contains/is about to find it instead of having to remember where you put it. (I personally dream of a day I can use such a thing in any file explorer anywhere, Godot, Windows itself, anything.)

I have personally worked on projects of various sizes that have used opposite approaches to file organization (by file type / by topic), and I have always found them to be equally frustrating and limiting in so many ways. 

Tags will go a long way towards rectifying this fundamental problem.

**Reply by pixel-boy** [reply]
I think you summed up the problem quite well: we're forced to choose a single approach when we'd like to have both at once.

In hindsight (and after testing the approach with a custom plugin) the tag system seems a bit heavy to me, but Obsidian's tagfolder system does indeed seem interesting.

Another approach I've thought of (and I wonder if the idea came from an existing plugin, but I can't find it anymore) that would be more independent could be an additional main screen (like 2D/3D/Script, etc.) allowing the creation of lists, into which files could be easily dragged and dropped, and which could be accessed via scripts.
Which is almost the same as a tag system.
By the way, right-clicking on files to assign them to lists could still be relevant with this approach.

Also being able to export list type variables (or a ProjectList resource ) which would allow assigning these lists via the inspector could be an interesting bonus,

I'm considering creating a list plugin as described above, but a native solution in Godot would make sense I think.

Somewhat related thought, the godot type "AudioStreamRandomizer" is already more or less a list, the current problem is that it is limited to sounds, why not have a TextureRandomizer? and why not have a generic concept that allows to create lists that can be assigned as resources, and that could manage the behavior of randomizer (and other type of selection based on parameters etc..)? It's a need that I often encounter, and even if coding it yourself is feasible, I would like a more solid solution, and easier to edit via an interface (global and local lists).
